tinygrad/codegen/kernel.py

```diff
@@ -704,12 +704,19 @@ def fix_st(warp_dims, tcd_dims, tcd_expand, pattern_1, pattern_2, st1):
           return LazyOp(op.op, (ret,), new_reduce_axes) if (new_reduce_axes:=tuple(i for i in arg if i-self.first_upcast not in reduce_axes)) else ret
         if self.group_for_reduces:
           start = LazyOp(op.op, tuple(fixup_ast(x, apply_to_st) for x in op.src), arg)
-          local_shape = (1,) * self.global_dims + self.full_shape[self.global_dims:self.global_dims+self.local_dims+self.group_for_reduces] + \
-            (1,) * (self.first_upcast - self.group_for_reduces - self.first_reduce) + tuple([x[0] for x in self.upcasted_axis(0)])
-          local_buffer = MemBuffer(-1, start.dtype, ShapeTracker.from_shape(local_shape))
+          arg = tuple(i for i in range(self.first_reduce, self.first_reduce+self.group_for_reduces) \
+                      if self.sts[reduce_idx].shape[i] != self.sts[reduce_idx+1].shape[i])
+          if len(arg) == 0: return start
+          local_shape = (1,) * self.global_dims + self.full_shape[self.global_dims:self.global_dims+self.local_dims] + \
+            tuple([self.full_shape[i] if self.sts[reduce_idx].shape[i] != self.sts[reduce_idx+1].shape[i] else 1 \
+              for i in range(self.first_reduce, self.first_reduce+self.group_for_reduces)]) + \
+            (1,) * (self.shape_len - self.upcasted - self.group_for_reduces - self.first_reduce) + tuple([x[0] for x in self.upcasted_axis(0)])
+          local_buffer = MemBuffer(-self.reduceops.index(op)-1, start.dtype, ShapeTracker.from_shape(local_shape))
```

Comment:
why do you need multiple temps for a back to back reduce?

Reply:
We could put it in one temp; that would just require an extra barrier. Would you prefer that?

Reply:
I've seen llm.c do the one buffer per reduceop, I think it's fine yea

Reply:
we'd need a way to do:
```
UOps.LOAD
UOps.BARRIER
UOps.STORE
```
whereas right now we only have a way to do
```
UOps.STORE
UOps.BARRIER
UOps.LOAD
```